### ISSUE TITLE
[Routeing] Add the option to avoid toll roads and unpaved roads

### ISSUE THREAD

**Issue by ignaciolep** [mapsme/omim#1888]
It would be useful to have the option to avoid toll and unpaved (dirt) roads.

**Comment by deathbaba** [comment]
It's in our plans, the task is not easy: current routing algorithm does not support it at all.

**Comment by ignaciolep** [comment]
I see. It's an OSRM limitation. Perhaps it's time to evaluate another open source routing engine that supports avoiding tolls like valhalla's toll_booth_penalty.

**Comment by RicoElectrico** [comment]
OSRM doesn't seem to be well-suited to mobile platforms anyway. The thing is, I don't know of any open source router that simply works (just like what commercial navigations have achieved like 10 years ago on hardware that was significantly less powerful).

**Comment by deathbaba** [comment]
CC @gardster @bykoianko

**Comment by PanierAvide** [comment]
Any news about this issue ? Is it still in your plans ?

**Comment by bykoianko** [comment]
Hi!
> Any news about this issue ? Is it still in your plans ?
> 

As far as I’m aware, yes it is still in our plans. There’s a limitation of contraction hierarchy (osrm). So implementing it is not easy.

Vladimir

**Comment by PanierAvide** [comment]
I don't see clearly the limitations you are talking about, OSRM seems to handle [several routing profiles](https://github.com/Project-OSRM/osrm-backend/tree/master/profiles), are there any issues to create a copy of car profile and restrict using roads with `toll=yes` tag ?

**Comment by deathbaba** [comment]
Current algorithm supports only one possible path for one profile. It means that to have "toll-free" routes, we need to build all routing data files separately and duplicate them (~ +30% to mwm file sizes).

**Comment by homersimpsons** [comment]
https://github.com/pgRouting/pgrouting but all engine need to be changed... Didn't check every capabilities

**Comment by homersimpsons** [comment]
Or graphhopper

**Comment by gardster** [comment]
@homersimpsons do you have the postgresql instance on your phone to use pgrouting? Graphhopper uses same technics with the OSRM and has same limitations.

**Comment by homersimpsons** [comment]
No, in fact I don't get postgre and porting it doesn't look easy/possible. But maybe you can try to adapt the code for SQLite ? 
In fact,  graph hopper also use CH, but others algorithms are implemented, like A* or Dijkstra, then maybe, for 'non CH' search, just display a reminder to prevent that it could be long.

**Comment by RicoElectrico** [comment]
I think you should check out [Thor](https://github.com/valhalla/thor) and help them port it to mobile.

**Comment by vvug** [comment]
Any news on this?

**Comment by burivuh** [comment]
We are changing algorithm right now, so you can expect your feature to be
implemented in the first half of 2017.
Sorry it is taking so long.

**Comment by vvug** [comment]
Great news, thank you!

**Comment by legovini** [comment]
Thanks for letting us know, @burivuh.
Is the new algorithm just a new version of OSRM, or something completely different?
In the latter case, which algorithm will maps.me use?

**Comment by kozuch** [comment]
Any news regarding implementation of this?

**Comment by ethomp10** [comment]
Bump… it’s 2018 now, any word? It’s a much needed feature IMO. :)

**Comment by burivuh** [comment]
Hi!
Bad news: we had to change our plan in 2017.
Good news: we have a green light for a car navigation development. At least for now.
Plan for the near future:
Improving/fixing turns generation - done for 8.2 (future major release)
Memory issues investigation - in the process
Car position extrapolation
Unpaved roads get an additional fine during route calculation
Unpaved and paid roads selection (mostly UI issue)

**Comment by ethomp10** [comment]
Cool, thanks for the update, keep up the great work!

**Comment by karlic** [comment]
I too would appreciate an "avoid highways" route. It would also be useful to say use only minor roads.

**Comment by ignaciolep** [comment]
Hi @burivuh,
It would be great if you could update us with the actual plan status.
Thanks!

> Hi!
> Bad news: we had to change our plan in 2017.
> Good news: we have a green light for a car navigation development. At least for now.
> Plan for the near future:
> Improving/fixing turns generation - done for 8.2 (future major release)
> Memory issues investigation - in the process
> Car position extrapolation
> Unpaved roads get an additional fine during route calculation
> Unpaved and paid roads selection (mostly UI issue)

**Comment by burivuh** [comment]
Improving/fixing turns generation - done for 8.2 (future major release)
Memory issues investigation - done for 8.2/8.3
Car position extrapolation - done for 8.3
Unpaved roads get an additional fine during route calculation - done for 8.3
Unpaved and paid roads selection (mostly UI issue) - in progress, plan to finish it this year (8.5/8.6)

Also, as we finally hire a full routing team:

City roads get an additional fine during out-city routes calculation - done for 8.4
Speed cam notifications - in progress for 8.5
Much better work with speed limits from OSM - in progress for 8.5
Time estimation tuning - in progress for 8.6

**Comment by ignaciolep** [comment]
Excellent. Thank you!!

**Comment by ignaciolep** [comment]
Hi @burivuh ,
As the lastest version in production is 8.6.3, Could you update us the tasks status? (Specially the "Unpaved and paid roads selection")
Thanks!

**Comment by ignaciolep** [comment]
It's seems that finally version 9.1.5 can avoid toll roads and unpaved roads! But it doesn't always work as expected.
Can anyone from staff explain how does it work the new router?
What are the exact conditions to avoid a toll road?
@burivuh @biodranik

**Comment by burivuh** [comment]
Hi!

If you select a setting to avoid some type of the road, router will not use that type for the route.

Have you found any bug or inconsistence behaviour? Please, send it to the support@maps.me so that we can properly track the issue.
